### DISCUSSION TITLE
max requests when not using worker mode

### DISCUSSION THREAD

**Discussion by fellmann** [php/frankenphp#1241] [Q&A] [answered]
Hi, we are using not worker mode as it is not supported by the application. I think the documentation could be improved for that case. Especially:
- How to you define the number of PHP threads that handle requests (num_threads)?
- Is there a configuration to restart threads after a certain amount of requests (like PM.max_requests in FPM)?
- Is there a configuration to restart threads after a certain time? For example to make sure persistent SQL connections are not used longer than a certain timed / killed after inactivity?
Thanks!

**Comment by dunglas** [comment]
Currently, threads never restart until you restart the whole server.

I'm not sure if there is a use case for restarting them.

**Comment by fellmann** [comment]
You are probably right. The manual says that it can be used to work around memory leaks, but this is hopefully not needed:
https://www.php.net/manual/en/install.fpm.configuration.php#pm.max-requests

But I am correct that num_threads is the equivalent to max_children in PHP-FPM with static mode?

**Reply by AlliBalliBaba** [reply]
Yeah you are right, it's similar when not using worker mode. The only difference is that it's using threads instead of forking processes.
A way to dynamically scale threads is on the roadmap.

After #1137, implementing a mechanic to respawn threads from time to time would be straightforward, but would it really be necessary? And would it actually circumvent leaks?

**Comment by fellmann** [comment]
Well, I think you are right, there is no practical use for it, only in case of memory leaks in PHP extensions it would make sense. And to avoid idle timeouts of database connections, which could also be solved by dynamic scaling.

Still, I am missing some documentation about running Frankenphp in non-worker mode (like using FPM), even if it is just one sentence. I might open a PR for that.

One more question, how does Frankenphp handle requests that exceed the thread limit (in non-worker mode)?
- Is there a limit how many requests can be waiting for an available PHP process?
- Is there a time limit how long requests can be waiting before an error is returned?

**Reply by dunglas** [reply, marked as answer]
> Still, I am missing some documentation about running Frankenphp in non-worker mode (like using FPM), even if it is just one sentence. I might open a PR for that.

Actually, it's the default. If you don't opt-in for the worker mode, you're in normal mode. PR welcome to make that more clear.


> Is there a limit how many requests can be waiting for an available PHP process?
> Is there a time limit how long requests can be waiting before an error is returned?

These limits apply if set https://caddyserver.com/docs/caddyfile/options#timeouts
I don't think we currently have a limit for the number of waiting requests, but we may add one.

**Reply by fellmann** [reply]
Can you point me to the code where incoming requests are queued until a PHP thread is available?

**Reply by AlliBalliBaba** [reply]
Non-worker requests are queued here:

https://github.com/dunglas/frankenphp/blob/d276032e20bd074f7d942194cfd25cf49474132f/frankenphp.go#L476-L480

I agree, there could be some form of additional limit there to help the server recover faster after being overburdened

**Reply by fellmann** [reply]
I could imagine a limit for the maximum number of requests waiting for a free thread, and a maximum waiting time. And in case of failure to respond in time, a 503 response would be great. 
Maybe even a Warning if requests are waiting because of not enough threads, like FPM does.
